### PR TITLE
Spatial_oM: Adding in inital 2D-layout objects

### LINKED ISSUE #758: Spatial_oM: Add Layout objects
#### Description:

As discussed in https://github.com/BHoM/BHoM/issues/579

Need to add 2D-layout objects, controlling distributions of points in a region/along a curve. To be used for Reinforcement (as mentioned) but can also be used for things like pile caps etc.

2 main things to be ironed out before implementing is (and related).

1. Should these be BHoMObjects or not
2. Which oM should they belong to. Geometry/Spatial feels like the most intuitive, but could be convinced otherwise. @al-fisher @pawelbaran @adecler @rwemay @FraserGreenroyd 

Proposed objects/interfaces:

```
    [Description("Base interface for all 2D layouts")]
    public interface ILayout2D : IBHoMObject
    {

    }
```

```
    [Description("Base interface for all 2D layouts")]
    public class PerimiterLayout : BHoMObject, ILayout2D
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        [Description("Number of points along the perimiter of the host object")]
        public int NumberOfPoints { get; set; }

        /***************************************************/
    }
```

```
    [Description("Layout for freeform control over the point distribution")]
    public class ExplicitLayout : BHoMObject, ILayout2D//, IImmutable //Maybe immutable, to make sure the points are always in xy-plane via create methods
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        [Description("The explicit grid of points in the layout")]
        public List<Point> Points { get; set; }

        /***************************************************/
    }
```

```
    [Description("Linear distribution of points along a vector from one side of the perimiter of the host object to the other.")]
    public class LinearLayout : BHoMObject, ILayout2D//, IImmutable //Maybe immutable, to make sure the points are always in xy-plane via create methods
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        [Description("Number of points along the axis.")]
        public int NumberOfPoints { get; set; }

        [Description("Direction of the axis.")]
        public Vector Direction { get; set; }

        [Description("Offset of the linear layout in relation to the centroid of the host object.")]
        public Vector Offset { get; set; }

        /***************************************************/
    }
```

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #758 

Adding initial 2D-layout objects. Initial use is for reinforcement, but intended to be usefull across more cases/disciplines as well.

Will create followup Engine PR that makes use of the classes

### CODE REVIEW

File: Spatial_oM/Layouts/LinearLayout.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        /***************************************************/
+
+        [Description("Number of points along the axis.")]
+        public int NumberOfPoints { get; }
+
+        [Description("Direction of the axis. Vector should lie in the XY-plane, i.e. have a Z-coordinate equal to 0.")]
+        public Vector Direction { get; }
+
+        [Description("Offset of the linear layout in relation to the centroid of the host object. Vector should lie in the XY-plane, i.e. have a Z-coordinate equal to 0.")]
+        public Vector Offset { get; }
```
Comment (kThorsager) [inline Spatial_oM/Layouts/LinearLayout.cs:48]: `Length` ?

File: Spatial_oM/Layouts/PerimiterLayout.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+
+using System.ComponentModel;
+using BH.oM.Base;
+
+namespace BH.oM.Spatial.Layouts
+{
+    [Description("Base interface for all 2D layouts")]
```
Comment (peterjamesnugent) [inline Spatial_oM/Layouts/PerimiterLayout.cs:29]: Linear distribution of points along the perimiter of the host object.

File: Spatial_oM/Layouts/PerimiterLayout.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("Number of points along the perimiter of the host object")]
+        public int NumberOfPoints { get; set; }
+
+        /***************************************************/
+    }
+}
```
Comment (peterjamesnugent) [inline Spatial_oM/Layouts/PerimiterLayout.cs:41]: Should we not have a constructor for a PerimeterLayout?

Comment (IsakNaslundBh) [reply Spatial_oM/Layouts/PerimiterLayout.cs:41]: Not needed, as not immutable. Only added it to the other ones as I wanted to have create methods for them making sure the input geometry was valid.

**Comment by al-fisher** [comment]
@IsakNaslundBh - for codeowners suggest same as Engine. 
@IsakNaslundBh @pawelbaran @al-fisher ?

**Comment by IsakNaslundBh** [comment]
> `PerimEterLayout`
> 
> Besides that, I would think of adding to `PerimeterLayout` a perimeter domain or any other way of defining layout only on part of the perimeter?

Hmm, not a bad idea. Wonder if it is the same object, or another one?

File: Spatial_oM/Layouts/LinearLayout.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        [Description("Number of points along the axis.")]
+        public virtual int NumberOfPoints { get; set; }
+
+        [Description("Direction of the axis. Vector should lie in the XY-plane, i.e. have a Z-coordinate equal to 0.")]
+        public virtual Vector Direction { get; }
+
+        [Length]
+        [Description("Offset of the linear layout in relation to the reference point, perpendicular to the Direction vector in the XY plane.")]
+        public virtual double Offset { get; }
```
Comment (kThorsager) [inline Spatial_oM/Layouts/LinearLayout.cs:50]: Should specify what positive values vs negative values imply

File: Spatial_oM/Layouts/MultiLinearLayout.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        [Length]
+        [Description("Minimum distance between any two layers.")]
+        public virtual double PerpendicularMinimumSpacing { get; set; }
+
+        [Description("Direction of the axis. Vector should lie in the XY-plane, i.e. have a Z-coordinate equal to 0.")]
+        public virtual Vector Direction { get; }
+
+        [Length]
+        [Description("Offset of the linear layout in relation to the reference point, perpendicular to the Direction vector in the XY plane.")]
+        public virtual double Offset { get; }
```
Comment (kThorsager) [inline Spatial_oM/Layouts/MultiLinearLayout.cs:59]: ^^

**Comment by IsakNaslundBh** [comment]
> Not sure, gut feeling is same object. Not sure, however, what would be best way of describing the domain to make it practical. I think it might make sense to first see how does the current PR work in practice and see if adding yet another layer of complexity is actually necessary.
>
> So good to merge on my end 👍

Cool, will fix the spelling stuff, and agree, this already grew a lot more than I first expected, and can always be added later!

File: Spatial_oM/Layouts/LinearLayout.cs (hunk 4 above)
Comment (IsakNaslundBh) [reply Spatial_oM/Layouts/LinearLayout.cs]: Good one. Done.

**Comment by IsakNaslundBh** [comment]
Had to update from master, as this is now lagging behind BHoM and BHoM_UI on master will not build without the changes.

**Comment by al-fisher** [comment]
@FraserGreenroyd @pawelbaran @peterjamesnugent looks like your requested changes have been addressed
@IsakNaslundBh is that right? 
Can we confirm ready for merge - thanks 😄

**Comment by al-fisher** [comment]
Thanks guys

**Pull request merged**